### DISCUSSION TITLE
support for flash write in RWW ?

### DISCUSSION THREAD

**Discussion by agsb** [Optiboot/optiboot#315] [Usage problems] [answered]
Hello,

With optiboot I can do flash writes at RWW ? There is any mode to do  for it ?

**Comment by WestfW** [comment, marked as answer]
I don't understand the question.
The DOSPM code will let you write anywhere in flash, but it always waits for the write to complete, even for the RWW section of flash.
The former "pipelined" code that would transfer the next page while a RWW page write was in progress was deleted a while ago to save space, after checking that it actually made very little difference in overall upload time.

**Comment by agsb** [comment]
Thanks for your response.

Let me try to explain.

From manuals, it is known that while program flash is not allowed to acess
(read/write) RWW, then the code for write flash must be at NRWW, buffer in
SRAM and parameters in registers.

Many implementors do not use optiboot because it do not leave the functions
for write flash "flexible" to be called from RWW, and return.

I did not know if the code of optiboot could be adapted to allow that
functionality, but it will be a great step for many implementations.

best regards,

Alvaro Barcellos

PS. English is not my first language.

On Tue, May 11, 2021 at 5:23 AM Bill Westfield ***@***.***>
wrote:

> I don't understand the question.
> The DOSPM code will let you write anywhere in flash, but it always waits
> for the write to complete, even for the RWW section of flash.
> The former "pipelined" code that would transfer the next page while a RWW
> page write was in progress was deleted a while ago to save space, after
> checking that it actually made very little difference in overall upload
> time.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Optiboot/optiboot/discussions/315#discussioncomment-722331>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR73TY2AV7RRH2UGC2OHHTTNDSOXANCNFSM44SNCPDQ>
> .
>


-- 
Slings move continents, a rock at a time.

**Comment by agsb** [comment]
I found this just now. My Fault!

"Now with do_spm() feature (allows applications to write to flash memory by
calling code in optiboot.)"

This should mean that this repository is now up-to-date with respect to
features and platforms added in the MCUDude mini/mega/mighty, and the
SpenceKonde Attiny cores.

thanks!


On Tue, May 11, 2021 at 9:27 AM Alvaro Barcellos ***@***.***>
wrote:

> Thanks for your response.
>
> Let me try to explain.
>
> From manuals, it is known that while program flash is not allowed to acess
> (read/write) RWW, then the code for write flash must be at NRWW, buffer in
> SRAM and parameters in registers.
>
> Many implementors do not use optiboot because it do not leave the
> functions for write flash "flexible" to be called from RWW, and return.
>
> I did not know if the code of optiboot could be adapted to allow that
> functionality, but it will be a great step for many implementations.
>
> best regards,
>
> Alvaro Barcellos
>
> PS. English is not my first language.
>
> On Tue, May 11, 2021 at 5:23 AM Bill Westfield ***@***.***>
> wrote:
>
>> I don't understand the question.
>> The DOSPM code will let you write anywhere in flash, but it always waits
>> for the write to complete, even for the RWW section of flash.
>> The former "pipelined" code that would transfer the next page while a RWW
>> page write was in progress was deleted a while ago to save space, after
>> checking that it actually made very little difference in overall upload
>> time.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Optiboot/optiboot/discussions/315#discussioncomment-722331>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ADR73TY2AV7RRH2UGC2OHHTTNDSOXANCNFSM44SNCPDQ>
>> .
>>
>
>
> --
> Slings move continents, a rock at a time.
>


-- 
Slings move continents, a rock at a time.

**Comment by agsb** [comment]
But where is that "do_spm()" ?

On Tue, May 11, 2021 at 12:07 PM Alvaro Barcellos <
***@***.***> wrote:

>
> I found this just now. My Fault!
>
> "Now with do_spm() feature (allows applications to write to flash memory
> by calling code in optiboot.)"
>
> This should mean that this repository is now up-to-date with respect to
> features and platforms added in the MCUDude mini/mega/mighty, and the
> SpenceKonde Attiny cores.
>
> thanks!
>
>
> On Tue, May 11, 2021 at 9:27 AM Alvaro Barcellos <
> ***@***.***> wrote:
>
>> Thanks for your response.
>>
>> Let me try to explain.
>>
>> From manuals, it is known that while program flash is not allowed to
>> acess (read/write) RWW, then the code for write flash must be at NRWW,
>> buffer in SRAM and parameters in registers.
>>
>> Many implementors do not use optiboot because it do not leave the
>> functions for write flash "flexible" to be called from RWW, and return.
>>
>> I did not know if the code of optiboot could be adapted to allow that
>> functionality, but it will be a great step for many implementations.
>>
>> best regards,
>>
>> Alvaro Barcellos
>>
>> PS. English is not my first language.
>>
>> On Tue, May 11, 2021 at 5:23 AM Bill Westfield ***@***.***>
>> wrote:
>>
>>> I don't understand the question.
>>> The DOSPM code will let you write anywhere in flash, but it always waits
>>> for the write to complete, even for the RWW section of flash.
>>> The former "pipelined" code that would transfer the next page while a
>>> RWW page write was in progress was deleted a while ago to save space, after
>>> checking that it actually made very little difference in overall upload
>>> time.
>>>
>>> —
>>> You are receiving this because you authored the thread.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/Optiboot/optiboot/discussions/315#discussioncomment-722331>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/ADR73TY2AV7RRH2UGC2OHHTTNDSOXANCNFSM44SNCPDQ>
>>> .
>>>
>>
>>
>> --
>> Slings move continents, a rock at a time.
>>
>
>
> --
> Slings move continents, a rock at a time.
>


-- 
Slings move continents, a rock at a time.

**Comment by WestfW** [comment]
there's an improved demo here: https://github.com/Optiboot/optiboot/tree/master/optiboot/examples/demo_flashwrite

>  this repository is now up-to-date with respect to features and platforms added in the MCUDude mini/mega/mighty, and the SpenceKonde Attiny cores.

Well, it has the ability to be called from an application to do flash writes via SPM.  But MCUDude and Spence have continued to add stuff, so we're still behind - no "flash page copy", for instance.

**Comment by agsb** [comment]
thanks!
again!

On Wed, May 12, 2021 at 5:10 AM Bill Westfield ***@***.***>
wrote:

> there's an improved demo here:
> https://github.com/Optiboot/optiboot/tree/master/optiboot/examples/demo_flashwrite
>
> this repository is now up-to-date with respect to features and platforms
> added in the MCUDude mini/mega/mighty, and the SpenceKonde Attiny cores.
>
> Well, it has the ability to be called from an application to do flash
> writes via SPM. But MCUDude and Spence have continued to add stuff, so
> we're still behind - no "flash page copy", for instance.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Optiboot/optiboot/discussions/315#discussioncomment-727564>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR73T3RW3M2NWMYK6XVWNLTNIZY7ANCNFSM44SNCPDQ>
> .
>


-- 
Slings move continents, a rock at a time.
